smart_open/gcs.py

```diff
@@ -0,0 +1,524 @@
+# -*- coding: utf-8 -*-
+"""Implements file-like objects for reading and writing to/from GCS."""
+
+import io
+import logging
+import sys
+
+from google.cloud import storage, exceptions
+import google.auth.transport.requests as google_requests
+import six
+
+import smart_open.bytebuffer
+
+logger = logging.getLogger(__name__)
+
+READ_BINARY = 'rb'
+WRITE_BINARY = 'wb'
+MODES = (READ_BINARY, WRITE_BINARY)
+"""Allowed I/O modes for working with GCS."""
+
+_BINARY_TYPES = (six.binary_type, bytearray)
+"""Allowed binary buffer types for writing to the underlying GCS stream"""
+if sys.version_info >= (2, 7):
+    _BINARY_TYPES = (six.binary_type, bytearray, memoryview)
+
+_UNKNOWN_FILE_SIZE = '*'
+
+_RESUME_INCOMPLETE = 308
+
+BINARY_NEWLINE = b'\n'
+
+SUPPORTED_SCHEMES = ("gcs", "gs")
+
+REQUIRED_CHUNK_MULTIPLE = 256 * 1024
+"""Google requires you to upload in multiples of 256 KB, except for the last part."""
+
+MIN_MIN_PART_SIZE = 256 * 1024
+"""The absolute minimum permitted by Google."""
+DEFAULT_BUFFER_SIZE = 256 * 1024
+
+START = 0
+CURRENT = 1
+END = 2
+WHENCE_CHOICES = [START, CURRENT, END]
+
+SUCCESSFUL_STATUS_CODES = (200, 201)
+
+
+def clamp(value, minval, maxval):
+    return max(min(value, maxval), minval)
+
+
+def make_range_string(start, stop=None, end=_UNKNOWN_FILE_SIZE):
+    #
+    # https://cloud.google.com/storage/docs/xml-api/resumable-upload#step_3upload_the_file_blocks
+    #
+    if stop is None:
+        return 'bytes %d-/%s' % (start, end)
+    return 'bytes %d-%d/%s' % (start, stop, end)
+
+
+class UploadFailedError(Exception):
+    """Raised when a multi-part upload to GCS returns a failed response status code."""
+
+
+def open(
+        bucket_id,
+        blob_id,
+        mode,
+        buffer_size=DEFAULT_BUFFER_SIZE,
+        min_part_size=MIN_MIN_PART_SIZE,
+        client=None,  # type: storage.Client
+        ):
+    """Open an GCS blob for reading or writing.
+
+    Parameters
+    ----------
+    bucket_id: str
+        The name of the bucket this object resides in.
+    blob_id: str
+        The name of the blob within the bucket.
+    mode: str
+        The mode for opening the object.  Must be either "rb" or "wb".
+    buffer_size: int, optional
+        The buffer size to use when performing I/O.
+    min_part_size: int, optional
+        The minimum part size for multipart uploads.  For writing only.
+    client: object, optional
```

Comment:
Be more precise about expected data types:

```suggestion
    client: google.cloud.storage.Client, optional
```